### ISSUE TITLE
Figure out how to best structure imports within the package

### ISSUE THREAD

**Issue by dschult** [networkx/networkx#4426]
This discussion crosses over a couple of PRs and shows we probably need to figure out how the imports are currently working and how we should best arrange them to work better.

I think they are better after #4401 #4403 and #4296. 
But the discussion in #4277 isn't conclusive and relative imports discussed in #4424 and  dschult#13 show this a bigger issue than we probably originally thought. 

Let's figure out the import order for the networkx namespace, propose a couple good ways to import networkx features from within the package and decide on one of them.

**Comment by dschult** [comment]
Nice thread from long long ago about relative imports.

https://mail.python.org/pipermail/python-dev/2003-December/041044.html

I can now see how relative imports could help if e.g. all scikits decided to merge into a single library.
Not fully understanding yet how that might impact us, or our users... still learning more...

**Comment by jarrodmillman** [comment]
So far the relative imports seem slightly worse than absolute imports (they are slightly more succinct, but make moving code more difficult).  I don't yet understand why we should avoid ``import networkx as nx`` in the library.  I still need to investigate where our circular imports are coming from.

**Comment by dschult** [comment]
Are we going to try to figure out import structure for v2.6?

If so, shift it back to milestone 2.6 and let's figure out what is causing the multiple/circular imports. Otherwise I have bumped the milestone to 3.0 and we'll try to get it right for that.

**Comment by rossbar** [comment]
I think getting rid of the erroneous exposures of the top-level namespace (e.g. `from networkx import nx`) *should* be straightforward: auditing the `__all__` attributes for the modules/pkgs should do it, though this might be pretty time consuming. It'd be nice to have for 2.6 but I don't think it's critical.

Changing the import style to use relative imports would probably also fix this issue, but would likely be more churn. IMO that feels like a 3.0 task, but perhaps only to me because I don't have a real sense of what should be done and how long it would take!

**Comment by dschult** [comment]
Nice -- let's try to fix the circular import stuff for 2.6 by auditing the ```__all__``` dunders.  Then we'll have a better idea of what's involved and can plan for what is 2.6 and 3.0 at that point.

**Comment by rossbar** [comment]
I think this issue has been largely addressed, at least in practice, with the advent of lazy imports and standardization of the import conventions (i.e. numpy as np, scipy as sp, etc.) in #7821.

The relative vs. absolute import discussion can be picked up again, but I think it fits best in a new, dedicated issue/discussion!
